pkg/doc/verifiable/json.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package verifiable
+
+import (
+	"encoding/json"
+)
+
+func marshalJSON(v interface{}, ef map[string]interface{}) ([]byte, error) {
+	// Convert value into a JSON map of known fields.
+	kf, err := valueToMap(v, ef)
+	if err != nil {
+		return nil, err
+	}
+
+	// Marshal extended known fields map.
+	return json.Marshal(kf)
+}
+
+func unmarshalJSON(data []byte, v interface{}) (map[string]interface{}, error) {
+	err := json.Unmarshal(data, v)
+	if err != nil {
+		return nil, err
+	}
+
+	// Collect known fields map.
+	vData, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+	var kf map[string]interface{}
+	err = json.Unmarshal(vData, &kf)
+	if err != nil {
+		return nil, err
+	}
+
+	// Collect all fields map.
+	ef := make(map[string]interface{})
+	err = json.Unmarshal(data, &ef)
+	if err != nil {
+		return nil, err
+	}
+
+	// Complement all fields with known fields to become a set of extra fields.
+	for k := range kf {
+		delete(ef, k)
+	}
+
+	return ef, nil
```

Comment:
it looks like this function returns unmarshalled extra fields in `data` that are not in `v`

retrieving these extra fields should be conveyed by the function name, `unmarshalJSON` is too broad

Reply:
Yeah, I will try to find a better name.

Reply:
Now it's `unmarshalValueWithMap()`.